File: src/jarvis/core/subagents/pool.py
"""SubagentPool — ThreadPoolExecutor-based parallel subagent execution.

Follows Hermes-Agent's delegate_task pattern: submit returns immediately,
results collected via drain-then-inject before each LLM call.
"""

from __future__ import annotations

from concurrent.futures import Future, ThreadPoolExecutor, TimeoutError as FutureTimeoutError
from threading import Lock
from typing import Any, Callable

from .models import SubagentConfig, SubagentHandle, SubagentStatus


class SubagentPool:
    """Manages parallel subagent execution via ThreadPoolExecutor."""

    def __init__(self, max_workers: int = 4, max_depth: int = 2) -> None:
        self._executor = ThreadPoolExecutor(max_workers=max_workers)
        self._agents: dict[str, SubagentHandle] = {}
        self._futures: dict[str, Future] = {}
        self._lock = Lock()
        self._notification_queue: list[dict[str, Any]] = []
        self.max_depth = max_depth
        self._run_fn: Callable | None = None  # set by AgentLoop after init

        # Metrics (align with Codex)
        self.spawn_count = 0
        self.completed_count = 0
        self.failed_count = 0
        self.total_steps = 0
        self.total_tokens = 0
        self.max_depth_reached = 0

    def set_runner(self, fn: Callable) -> None:
        """Set the function used to run a subagent in a thread."""
        self._run_fn = fn

    def submit(self, config: SubagentConfig) -> SubagentHandle:
        """Submit a subagent for async execution. Returns immediately."""
        if self._run_fn is None:
            raise RuntimeError("SubagentPool.run_fn not set -- call set_runner() first")

        if config.depth > self.max_depth:
            return SubagentHandle(
                agent_id=config.agent_id,
                agent_type=config.agent_type,
                status=SubagentStatus.FAILED,
                error=f"Depth {config.depth} exceeds max {self.max_depth}",
                depth=config.depth,
            )

        handle = SubagentHandle(
            agent_id=config.agent_id,
            agent_type=config.agent_type,
            status=SubagentStatus.RUNNING,
            max_steps=config.budget_steps,
            depth=config.depth,
        )

        future = self._executor.submit(self._run_wrapped, config, handle)
        with self._lock:
            self._agents[config.agent_id] = handle
            self._futures[config.agent_id] = future
            self.spawn_count += 1
            self.max_depth_reached = max(self.max_depth_reached, config.depth)

        return handle

    def _run_wrapped(self, config: SubagentConfig, handle: SubagentHandle) -> None:
        try:
            result = self._run_fn(config)
            with self._lock:
                handle.status = SubagentStatus.COMPLETED
                handle.result = str(result.get("final_answer", ""))[:8000]
                handle.steps = int(result.get("steps", 0))
                handle.total_tokens = int(result.get("total_tokens", 0))
                self.completed_count += 1
                self.total_steps += handle.steps
                self.total_tokens += handle.total_tokens
                self._notification_queue.append({
                    "agent_id": config.agent_id,
                    "agent_type": config.agent_type,
                    "status": "completed",
                    "result": handle.result,
                    "steps": handle.steps,
                    "total_tokens": handle.total_tokens,
                })
        except Exception as exc:
            with self._lock:
                handle.status = SubagentStatus.FAILED
                handle.error = str(exc)
                self.failed_count += 1
                self._notification_queue.append({
                    "agent_id": config.agent_id,
                    "agent_type": config.agent_type,
                    "status": "failed",
                    "error": str(exc),
                })

    def list_agents(self) -> list[dict[str, Any]]:
        """Return status of all agents (active and completed)."""
        with self._lock:
            return [
                {
                    "agent_id": h.agent_id,
                    "agent_type": h.agent_type,
                    "status": h.status.value,
                    "steps": h.steps,
                    "max_steps": h.max_steps,
                    "depth": h.depth,
                    "error": h.error,
                }
                for h in self._agents.values()
            ]

    def wait_agent(self, agent_id: str, timeout: float = 60.0) -> dict[str, Any]:
        """Block until a specific agent completes or timeout expires."""
        with self._lock:
            handle = self._agents.get(agent_id)
            future = self._futures.get(agent_id)
            if handle is None:
                return {"agent_id": agent_id, "status": "not_found"}
            status = handle.status
            result = handle.result
            error = handle.error
        if status in (SubagentStatus.COMPLETED, SubagentStatus.FAILED, SubagentStatus.CANCELLED):
            return {"agent_id": agent_id, "status": status.value, "result": result, "error": error}
        if future is not None:
            try:
                future.result(timeout=timeout)
            except FutureTimeoutError:
                pass
        with self._lock:
            h = self._agents.get(agent_id)
            if h is None:
                return {"agent_id": agent_id, "status": "not_found"}
            return {"agent_id": agent_id, "status": h.status.value, "result": h.result, "error": h.error}

    def close_agent(self, agent_id: str) -> bool:
        """Cancel a running agent. Returns True if cancelled."""
        with self._lock:
            handle = self._agents.get(agent_id)
            future = self._futures.get(agent_id)
            if handle is None:
                return False
            status = handle.status
        if status != SubagentStatus.RUNNING:
            return False
        if future is not None and not future.done():
            future.cancel()
        with self._lock:
            handle.status = SubagentStatus.CANCELLED
        return True

    def active_count(self) -> int:
        with self._lock:
            return sum(1 for h in self._agents.values() if h.status == SubagentStatus.RUNNING)

    def drain_notifications(self) -> list[dict[str, Any]]:
        """Return and clear all pending completion notifications."""
        with self._lock:
            notifs = list(self._notification_queue)
            self._notification_queue.clear()
        return notifs

    def shutdown(self, *, wait: bool = True) -> None:
        self._executor.shutdown(wait=wait)

File: packages/subagents/src/pool.ts
// ============================================================================
// SubagentPool — manage concurrent subagent execution
// ============================================================================

import type { SubagentConfig, SubagentHandle, SubagentResult, SubagentStatus } from './models.js';
import { AgentMailbox } from '@jarvis/agent';

// ============================================================================
// SubagentPool
// ============================================================================

export class SubagentPool {
  private agents: Map<string, SubagentHandle> = new Map();
  private mailboxes: Map<string, AgentMailbox> = new Map();
  private runner: ((config: SubagentConfig, mailbox: AgentMailbox) => Promise<SubagentResult>) | null = null;
  private notifications: Array<{ agentId: string; status: SubagentStatus; result?: SubagentResult }> = [];
  private activeCount = 0;
  private pending: Array<() => void> = [];

  /** Maximum concurrent subagents (default 4, matches Codex agent_max_threads). */
  readonly maxConcurrent: number;

  /** Default nesting depth for subagents. Set before submitting. */
  defaultDepth = 0;

  /** Parent mailbox — subagent results auto-delivered here on completion. */
  private parentMailbox: AgentMailbox | null = null;

  constructor(maxConcurrent = 4) {
    this.maxConcurrent = Math.max(1, maxConcurrent);
  }

  /**
   * Set the parent mailbox where all subagent results are auto-delivered.
   * This enables asynchronous spawn: parent spawns → returns immediately →
   * results arrive automatically in parent's mailbox.
   */
  setParentMailbox(mailbox: AgentMailbox): void {
    this.parentMailbox = mailbox;
  }

  // ========================================================================
  // Configuration
  // ========================================================================

  /**
   * Set the runner function that executes subagent tasks.
   * Called once before any subagents are submitted.
   */
  setRunner(fn: (config: SubagentConfig, mailbox: AgentMailbox) => Promise<SubagentResult>): void {
    this.runner = fn;
  }

  /** Get the mailbox for a specific subagent. */
  getMailbox(agentId: string): AgentMailbox | undefined {
    return this.mailboxes.get(agentId);
  }

  // ========================================================================
  // Submit
  // ========================================================================

  /**
   * Submit a subagent for execution.
   * Returns a handle immediately. The subagent runs asynchronously.
   * Multiple subagents run concurrently up to maxConcurrent.
   */
  submit(config: SubagentConfig): SubagentHandle {
    if (!this.runner) {
      throw new Error('SubagentPool: no runner configured. Call setRunner() first.');
    }

    if (this.agents.has(config.agentId)) {
      throw new Error(`Subagent ${config.agentId} already exists`);
    }

    let resolveCompletion: (result: SubagentResult) => void;
    const completion = new Promise<SubagentResult>((resolve) => {
      resolveCompletion = resolve;
    });

    let cancelled = false;

    const handle: SubagentHandle = {
      agentId: config.agentId,
      status: 'pending',
      completion,
      cancel: () => {
        cancelled = true;
        this._updateStatus(config.agentId, 'cancelled', {
          agentId: config.agentId,
          status: 'cancelled',
          error: 'Cancelled',
        });
        resolveCompletion!({
          agentId: config.agentId,
          status: 'cancelled',
          error: 'Cancelled',
        });
      },
    };

    this.agents.set(config.agentId, handle);

    // Create a dedicated mailbox for this subagent
    const mailbox = new AgentMailbox();
    this.mailboxes.set(config.agentId, mailbox);

    // Start async execution (respects maxConcurrent)
    this._execute(config, handle, mailbox, resolveCompletion!, cancelled);

    return handle;
  }

  // ========================================================================
  // Query
  // ========================================================================

  /** List all subagents with their current status. */
  listAgents(): Array<{ agentId: string; status: SubagentStatus }> {
    return [...this.agents.values()].map((h) => ({
      agentId: h.agentId,
      status: h.status,
    }));
  }

  /** Wait for a specific subagent to complete. */
  async waitAgent(agentId: string, timeoutMs?: number): Promise<SubagentResult> {
    const handle = this.agents.get(agentId);
    if (!handle) {
      throw new Error(`Subagent ${agentId} not found`);
    }

    if (timeoutMs) {
      const timeout = new Promise<SubagentResult>((_, reject) =>
        setTimeout(() => reject(new Error(`Timeout waiting for ${agentId}`)), timeoutMs),
      );
      return Promise.race([handle.completion, timeout]);
    }

    return handle.completion;
  }

  /** Get the number of currently active (pending/running) subagents. */
  getActiveCount(): number {
    let count = 0;
    for (const h of this.agents.values()) {
      if (h.status === 'pending' || h.status === 'running') count++;
    }
    return count;
  }

  /** Drain any accumulated notifications. */
  drainNotifications(): Array<{
    agentId: string;
    status: SubagentStatus;
    result?: SubagentResult;
  }> {
    const drained = [...this.notifications];
    this.notifications = [];
    return drained;
  }

  /** Remove all completed/failed/cancelled agents and shut down. */
  shutdown(): void {
    for (const [, handle] of this.agents) {
      if (handle.status === 'pending' || handle.status === 'running') {
        handle.cancel();
      }
    }
    this.agents.clear();
    this.mailboxes.clear();
    this.notifications = [];
    this.pending = [];
    this.activeCount = 0;
  }

  // ========================================================================
  // Internal
  // ========================================================================

  private async _acquireSlot(): Promise<void> {
    if (this.activeCount < this.maxConcurrent) {
      this.activeCount++;
      return;
    }

    // Queue up — wait for a slot to open
    return new Promise<void>((resolve) => {
      this.pending.push(() => {
        this.activeCount++;
        resolve();
      });
    });
  }

  private _releaseSlot(): void {
    this.activeCount--;
    // Unblock next pending subagent
    const next = this.pending.shift();
    if (next) {
      // Release on next microtick so stack doesn't grow unbounded
      setImmediate(next);
    }
  }

  private async _execute(
    config: SubagentConfig,
    handle: SubagentHandle,
    mailbox: AgentMailbox,
    resolve: (result: SubagentResult) => void,
    cancelled: boolean,
  ): Promise<void> {
    // Wait for concurrency slot (non-blocking)
    await this._acquireSlot();

    if (cancelled) {
      this._releaseSlot();
      return;
    }

    this._updateStatus(config.agentId, 'running');

    try {
      const result = await this.runner!(config, mailbox);
      if (!cancelled) {
        this._updateStatus(config.agentId, result.status, result);
        resolve(result);

        // Auto-deliver result to parent mailbox (Codex completion watcher pattern)
        if (this.parentMailbox) {
          const summary = result.answer || result.error || '(no output)';
          this.parentMailbox.deliver(
            config.agentId,
            `[Subagent ${result.status}]\nTask: ${config.task.slice(0, 100)}\nResult: ${summary.slice(0, 500)}`,
            true,
          );
        }
      }
    } catch (err) {
      const errMsg = err instanceof Error ? err.message : String(err);
      const failedResult: SubagentResult = {
        agentId: config.agentId,
        status: 'failed',
        error: errMsg,
      };
      if (!cancelled) {
        this._updateStatus(config.agentId, 'failed', failedResult);
        resolve(failedResult);
      }
    } finally {
      this.mailboxes.delete(config.agentId);
      this._releaseSlot();
    }
  }

  /** External callback for TUI agent store updates. */
  onStatusUpdate?: (entry: { agentId: string; status: string; role?: string; depth?: number; task?: string }) => void;

  private _updateStatus(
    agentId: string,
    status: SubagentStatus,
    result?: SubagentResult,
  ): void {
    const handle = this.agents.get(agentId);
    if (handle) {
      (handle as { status: SubagentStatus }).status = status;
    }
    this.notifications.push({ agentId, status, result });
    this.onStatusUpdate?.({ agentId, status });
  }
}